src/main/java/dan200/computercraft/client/render/MonitorShader.java

```diff
@@ -0,0 +1,155 @@
+        transform.write( MATRIX_BUFFER );
+        MATRIX_BUFFER.rewind();
+        RenderSystem.glUniformMatrix4( UNIFORM_MV, false, MATRIX_BUFFER );
+
+        // TODO: Cache this?
+        MATRIX_BUFFER.rewind();
+        GL11.glGetFloatv( GL11.GL_PROJECTION_MATRIX, MATRIX_BUFFER );
+        MATRIX_BUFFER.rewind();
+        RenderSystem.glUniformMatrix4( UNIFORM_P, false, MATRIX_BUFFER );
+
+        RenderSystem.glUniform1i( UNIFORM_WIDTH, width );
+        RenderSystem.glUniform1i( UNIFORM_HEIGHT, height );
+
+        // TODO: Cache this? Maybe??
+        PALETTE_BUFFER.rewind();
+        for( int i = 0; i < 16; i++ )
+        {
+            double[] colour = palette.getColour( i );
+            if( greyscale )
+            {
+                float f = FixedWidthFontRenderer.toGreyscale( colour );
+                PALETTE_BUFFER.put( f ).put( f ).put( f );
+            }
+            else
+            {
+                PALETTE_BUFFER.put( (float) colour[0] ).put( (float) colour[1] ).put( (float) colour[2] );
+            }
+        }
+        PALETTE_BUFFER.flip();
+        RenderSystem.glUniform3( UNIFORM_PALETTE, PALETTE_BUFFER );
+    }
+
+    static boolean use()
+    {
+        if( initialised )
+        {
+            if( ok ) GlStateManager.useProgram( program );
+            return ok;
+        }
+
+        if( ok = load() )
+        {
+            GL20.glUseProgram( program );
+            RenderSystem.glUniform1i( MonitorShader.UNIFORM_FONT, 0 );
+            RenderSystem.glUniform1i( MonitorShader.UNIFORM_TBO, TEXTURE_INDEX - GL13.GL_TEXTURE0 );
+        }
+
+        return ok;
+    }
+
+    private static boolean load()
+    {
+        initialised = true;
+
+        try
+        {
+            int vertexShader = loadShader( GL20.GL_VERTEX_SHADER, "assets/computercraft/shaders/monitor.vert" );
+            int fragmentShader = loadShader( GL20.GL_FRAGMENT_SHADER, "assets/computercraft/shaders/monitor.frag" );
+
+            program = GlStateManager.createProgram();
+            GlStateManager.attachShader( program, vertexShader );
+            GlStateManager.attachShader( program, fragmentShader );
+            GlStateManager.linkProgram( program );
+
+            int i = GlStateManager.getProgram( program, GL20.GL_LINK_STATUS );
+            if( i == 0 )
+            {
+                ComputerCraft.log.warn( "Error encountered when linking monitor shaders." );
+                ComputerCraft.log.warn( GlStateManager.getProgramInfoLog( program, Short.MAX_VALUE ) );
+            }
+
+            return true;
+        }
+        catch( Exception e )
+        {
+            ComputerCraft.log.error( "Cannot load monitor shaders", e );
+            return false;
+        }
+
+        /*
+        GL20.glDetachShader( vertexShader, program );
+        GL20.glDetachShader( fragmentShader, program );
+        GlStateManager.deleteShader( vertexShader );
+        GlStateManager.deleteShader( fragmentShader );
```

Comment:
Not sure why you commented this out. You can and should delete shaders after they've been linked into a program.

Reply:
I was at the stage of "comment everything out until it works". Then forgot to add them back. Thanks!.